joint_limits/include/joint_limits/joint_limiter_interface.hpp

```diff
@@ -0,0 +1,181 @@
+      if (!declare_parameters(joint_names[i], node_param_itf_, node_logging_itf_))
+      {
+        RCLCPP_ERROR(
+          node_logging_itf_->get_logger(),
+          "JointLimiter: Joint '%s': parameter declaration has failed", joint_names[i].c_str());
+        result = false;
+        break;
+      }
+      if (!get_joint_limits(joint_names[i], node_param_itf_, node_logging_itf_, joint_limits_[i]))
+      {
+        RCLCPP_ERROR(
+          node_logging_itf_->get_logger(),
+          "JointLimiter: Joint '%s': getting parameters has failed", joint_names[i].c_str());
+        result = false;
+        break;
+      }
+      RCLCPP_INFO(
+        node_logging_itf_->get_logger(), "Limits for joint %zu (%s) are:\n%s", i,
+        joint_names[i].c_str(), joint_limits_[i].to_string().c_str());
+    }
+
+    if (result)
+    {
+      result = on_init();
+    }
+
+    // avoid linters output
+    (void)robot_description_topic;
+
+    return result;
+  }
+
+  /**
+   * Wrapper init method that accepts pointer to the Node.
+   * For details see other init method.
+   */
+  JOINT_LIMITS_PUBLIC virtual bool init(
+    const std::vector<std::string> & joint_names, const rclcpp::Node::SharedPtr & node,
+    const std::string & robot_description_topic = "/robot_description")
+  {
+    return init(
+      joint_names, node->get_node_parameters_interface(), node->get_node_logging_interface(),
+      robot_description_topic);
+  }
+
+  /**
+   * Wrapper init method that accepts pointer to the LifecycleNode.
+   * For details see other init method.
+   */
+  JOINT_LIMITS_PUBLIC virtual bool init(
+    const std::vector<std::string> & joint_names,
+    const rclcpp_lifecycle::LifecycleNode::SharedPtr & lifecycle_node,
+    const std::string & robot_description_topic = "/robot_description")
+  {
+    return init(
+      joint_names, lifecycle_node->get_node_parameters_interface(),
+      lifecycle_node->get_node_logging_interface(), robot_description_topic);
+  }
+
+  JOINT_LIMITS_PUBLIC virtual bool configure(
+    const trajectory_msgs::msg::JointTrajectoryPoint & current_joint_states)
+  {
+    // TODO(destogl): add checks for position
+    return on_configure(current_joint_states);
+  }
+
+  JOINT_LIMITS_PUBLIC virtual bool enforce(
+    trajectory_msgs::msg::JointTrajectoryPoint & current_joint_states,
+    trajectory_msgs::msg::JointTrajectoryPoint & desired_joint_states, const rclcpp::Duration & dt)
+  {
+    // TODO(destogl): add checks if sizes of vectors and number of limits correspond.
+    return on_enforce(current_joint_states, desired_joint_states, dt);
+  }
+
+protected:
+  // Methods that each limiter implementation has to implement
+  JOINT_LIMITS_PUBLIC virtual bool on_init() { return true; }
+
+  JOINT_LIMITS_PUBLIC virtual bool on_configure(
+    const trajectory_msgs::msg::JointTrajectoryPoint & /*current_joint_states*/)
+  {
+    return true;
+  }
```

Comment:
If on_init and on_configure are things that each limiter has to implement, why not make them pure
virtual?. Atleast on_init make sense to be enforced right?.

What do you think?

Reply:
That's good. Let me formulate it this way, Why not have them as pure virtual altogether and return as true from inside the `SimpleJointLimiter`. In that way, we might force everyone to declare them and initialize them. 

This is just an implementation detail. Just a minor comment

Reply:
Agree, adjusted!